README.md
#TBIB Loading Transition button

A Customizable transition button for Flutter

## Getting Started

To use this package, add `tbib_loading_transition_button_and_social` as a dependency in your `pubspec.yaml` file.

```yaml
dependencies:
 ...
 tbib_loading_transition_button_and_social: ^1.0.0
```

## How to use

In your project add the following import:

```dart
import  'package:tbib_loading_transition_button_and_social/tbib_loading_transition_button_and_social.dart';
```

In order to use this widget you have to use a `LoadingButtonController` to handler the different states.

```dart
final _controller = LoadingButtonController();
```

This widget starts a loading animation once it's tapped or by using the controller. You can the launcher to
init an error animation or stop the loading animation.

```dart
LoadingButton(
    color: Colors.blue,
    onSubmit: () => print('onSubmit'),
    controller: _controller,
    errorColor: Colors.red,
    transitionDuration: Duration(seconds: 1),
    child: Text(
    'Hit me!',
    style: Theme.of(context)
        .textTheme
        .bodyText1!
        .copyWith(color: Colors.white),
    ),
),
```

To support the transition to a different page to have to call the `moveToScreen` method

```dart
_controller.moveToScreen(
    context: context,
    page: SearchPage(),
    stopAnimation: true,
    navigationCallback: (route) =>
        Navigator.of(context).push(route),
),
```

// available widget success

```dart
LoadingButton(
    // v1.0.1
    borderSize:8
    boarderSide: BoarderSide.none, // Default
    color: Colors.blue,
    onSubmit: () => print('onSubmit'),
    controller: _controller,
    errorColor: Colors.red,
    progressIndicatorColor: Colors.grey,
    durationSuccess: const Duration(seconds: 1),
    duration: const Duration(milliseconds: 500),
    successWidget: FaIcon(FontAwesomeIcons.checkCircle),
    transitionDuration: Duration(seconds: 1),
    child: Text(
    'Hit me!',
    style: Theme.of(context)
        .textTheme
        .bodyText1!
        .copyWith(color: Colors.white),
    ),
),
```


## Login Button


Now in your Dart code, you can use:


## Usage Example

It very simple!

In order to use this widget you have to use a `LoadingSignButtonController` to handler the different states.

```dart
final _controller = LoadingSignButtonController();
```

```dart
LoadingSignButton(
  controller: _controller,
  buttonType: ButtonType.google,

  onPressed: () {
   print('click');
  })
```


* ButtonSize Removed in v 1.0.0
```dart
LoadingSignButton(
  controller: _controller,
  buttonType: ButtonType.google,
  buttonSize: ButtonSize.large, // small(default), medium, large
  onPressed: () {
   print('click');
  })
```

* use it in v 1.0.0
```dart
LoadingSignButton(
  controller: _controller,
  buttonType: ButtonType.google,
  width: 100,
  height:50,
  fontSize: 20,
  imageSize: 16,
  onPressed: () {
   print('click');
  })
```

* ImagePosition
```dart
LoadingSignButton(
  controller: _controller,
  imagePosition: ImagePosition.left, // left or right
  buttonType: ButtonType.google,
  onPressed: () {
   print('click');
  })
```

* Customized Button
```dart
LoadingSignButton(
controller: _controller,
 buttonType: ButtonType.pinterest,
 imagePosition: ImagePosition.right,
 //[buttonSize] You can also use this in combination with [width]. Increases the font and icon size of the button.
 buttonSize: ButtonSize.large,
 btnTextColor: Colors.grey,
 btnColor: Colors.white,
 //[width] Use it to change width button.
  //[height] Use it to change height button.

 btnText: 'Pinterest',
 onPressed: () {
  print('click');
 })
```


* Disabled Button

```dart
LoadingSignButton(
 controller: _controller,
 buttonType: ButtonType.facebook,
 onPressed: null,
),
```



### Mini Button
```dart
LoadingSignButton.mini(
 buttonType: ButtonType.github,
 onPressed: () {},
),
```

# TBIB Loading Transition button

![AnimatedButton](https://raw.githubusercontent.com/the-best-is-best/tbib_loading_transition_button_and_social/master/src/example1.gif)


CHANGELOG.md
## 1.0.1

* now you can custom BorderRadiusSize && fixed bugs

## 1.0.0

* buttonSize Remived and can add with, height , imageSize and fontSize in buttonSignSocial 
* support flutter 3 

## 0.0.4

* add video in readme.

## 0.0.2 - 0.0.3

* Update Read me Sign In Animation.

## 0.0.1

* TODO: Describe initial release.


lib/src/loading_button.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:vector_math/vector_math_64.dart' as v;

class LoadingButton extends StatefulWidget {
  /// Background color of the button for the default / steady state.  If not set
  /// this will default to the accent color of the current [Theme].
  final Color? color;

  /// Controller to be able to notify when the button can
  /// move to the next page o should animate to an
  /// errored state
  final LoadingButtonController controller;

  /// Duration for the button animation.
  final Duration duration;

  /// Duration for the page trnasition
  final Duration transitionDuration;

  /// Background color for error state
  /// if null, it will use [color] by default
  final Color errorColor;

  // new
  ///  widget success for error state
  final Widget successWidget;

  // new
  /// Duration for the  success button animation.
  final Duration durationSuccess;

  // new
  final double height;

  //new
  final double width;

  /// Use this callback to listen for interactions
  /// with the loading button
  final VoidCallback? onSubmit;

  /// The child widget for the loading button
  final Widget child;

  final Color progressIndicatorColor;
  final double? borderSize;
  final BorderSide? borderSide;

  /// A button that respresents a loading and errored stated
  /// if given a [LoadingButtonController], this widget can use the [moveToScreen]
  /// to animate a transition to a new page
  LoadingButton({
    Key? key,
    this.color,
    this.duration = const Duration(milliseconds: 300),
    this.transitionDuration = const Duration(milliseconds: 400),
    this.onSubmit,
    required this.controller,
    required this.child,
    this.errorColor = Colors.red,
    this.width = 100,
    this.height = 30,
    this.progressIndicatorColor = Colors.black,
    //new
    this.successWidget =
        const Icon(Icons.check_box, size: 16, color: Colors.green),
    this.durationSuccess = const Duration(milliseconds: 1000),
    this.borderSize = 8,
    this.borderSide,
  })  : assert(duration.inMilliseconds > 0),
        super(key: key);

  @override
  // ignore: library_private_types_in_public_api
  _LoadginButtonState createState() => _LoadginButtonState();
}

class _LoadginButtonState extends State<LoadingButton>
    with TickerProviderStateMixin, RouteAware {
  late AnimationController _controller;

  /// if null, it will use the accent color by default
  Color? _buttonColor;
  bool _isLoading = false;
  bool _isError = false;
  bool _isReversed = false;
  double _parentHeight = 0;

  //new
  bool _isSuccess = false;
  late Duration _durationSuccess;

  // Transition
  final GlobalKey _buttonKey = GlobalKey();

  final RouteObserver<PageRoute<dynamic>> routeObserver =
      RouteObserver<PageRoute>();

  @override
  void initState() {
    widget.controller._state = this;

    _controller = AnimationController(
      vsync: this,
      duration: widget.duration,
    )..addListener(() => setState(() {}));

    super.initState();
  }

  @override
  void didUpdateWidget(LoadingButton oldWidget) {
    super.didUpdateWidget(oldWidget);
    _durationSuccess = widget.durationSuccess;
    _buttonColor = widget.color ?? Theme.of(context).colorScheme.secondary;
    routeObserver.subscribe(this, ModalRoute.of(context) as PageRoute<dynamic>);
  }

  @override
  didPopNext() {
    setState(() => _isLoading = false);
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _durationSuccess = widget.durationSuccess;
    _buttonColor = widget.color ?? Theme.of(context).colorScheme.secondary;
    routeObserver.subscribe(this, ModalRoute.of(context) as PageRoute<dynamic>);

    Animation startColor = ColorTween(
      begin: widget.color,
      end: widget.errorColor,
    ).animate(
      CurvedAnimation(
        parent: _controller,
        curve: const Interval(0.0, 0.2),
      ),
    );

    startColor.addListener(() {
      _buttonColor = startColor.value;
    });

    Animation endColor = ColorTween(
      begin: widget.errorColor,
      end: widget.color,
    ).animate(
      CurvedAnimation(
        parent: _controller,
        curve: const Interval(0.8, 1.0),
      ),
    );

    endColor.addListener(() {
      _buttonColor = endColor.value;
    });
  }

  @override
  void dispose() {
    _controller.dispose();
    //routeObserver.unsubscribe(this);

    super.dispose();
  }

  // Shake animation
  v.Vector3 _getTranslation() {
    var progress = _controller.value;
    var offset = sin(progress * pi * 5);

    offset *= 12;
    return v.Vector3(offset, 0.0, 0.0);
  }

  Widget _getChildWidget(BuildContext context) {
    return AnimatedSwitcher(
      switchOutCurve: Curves.easeInOutCubic,
      duration: widget.duration,
      child: _isLoading && !_isSuccess
          ? CircularProgressIndicator(
              strokeWidth: 2.0,
              valueColor: AlwaysStoppedAnimation(widget.progressIndicatorColor),
            )
          : _isLoading && _isSuccess
              ? widget.successWidget
              : widget.child,
    );
  }

  double _getBeginWidth(BoxConstraints constraints) {
    if (_isReversed) {
      return _parentHeight;
    }
    return constraints.maxWidth;
  }

  double _getEndWidth(BoxConstraints constraints) {
    if (_isLoading) {
      return _parentHeight;
    }
    return constraints.maxWidth;
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: widget.height,
      width: widget.width,
      child: LayoutBuilder(
          builder: (BuildContext context, BoxConstraints constraints) {
        _parentHeight = constraints.maxHeight;
        return Center(
          child: TweenAnimationBuilder(
            curve: Curves.fastOutSlowIn,
            duration: widget.duration,
            tween: Tween<double>(
                begin: _getBeginWidth(constraints),
                end: _getEndWidth(constraints)),
            child: Transform(
              transform: Matrix4.translation(_getTranslation()),
              child: RawMaterialButton(
                key: _buttonKey,
                onPressed: () => (_isLoading || widget.onSubmit == null)
                    ? null
                    : _submitAndAnimate(),
                fillColor: _buttonColor,
                shape: RoundedRectangleBorder(
                  side: widget.borderSide ?? BorderSide.none,
                  borderRadius: BorderRadius.circular(
                      widget.borderSize ?? constraints.maxHeight / 2),
                ),
                child: _getChildWidget(context),
              ),
            ),
            builder: (BuildContext context, double size, Widget? child) {
              return SizedBox(
                height: constraints.maxHeight,
                width: size,
                child: child,
              );
            },
            onEnd: () {
              if (_isError) {
                _controller.reset();
                _controller.forward();
                _isError = false;
              }
              if (_isSuccess) {
                _isSuccess = false;
              }
            },
          ),
        );
      }),
    );
  }

  /// Calls [onSubmit] callback and inits the animation
  void _submitAndAnimate() {
    widget.onSubmit?.call();
    _iniAnimation();
  }

  void _stopAnimation() {
    setState(() {
      _isReversed = true;
      _isLoading = false;
    });
  }

  void _iniAnimation() {
    setState(() {
      _isReversed = false;
      _isError = false;
      _isLoading = true;
      _isSuccess = false;
    });
  }

  void _success() {
    setState(() {
      _isSuccess = true;
    });
  }

  Widget _buildTransition(
    BuildContext context,
    Widget page,
    Animation<double> animation,
    Size buttonSize,
    Offset fabOffset,
  ) {
    if (animation.value == 1) return page;

    final rect = Rect.fromCenter(
        center: fabOffset, width: buttonSize.width, height: buttonSize.height);
    final finalRect = rect.inflate(MediaQuery.of(context).size.longestSide);

    final radiusTween = BorderRadiusTween(
      begin: BorderRadius.circular(widget.borderSize ?? buttonSize.height / 2),
      end:
          BorderRadius.circular(widget.borderSize ?? finalRect.size.height / 2),
    );

    final sizeTween = SizeTween(
      begin: buttonSize,
      end: finalRect.size,
    );

    final offsetTween = Tween<Offset>(
        begin: fabOffset,
        end: Offset(MediaQuery.of(context).size.width - finalRect.right,
            MediaQuery.of(context).size.height - finalRect.bottom));

    final easeInAnimation = CurvedAnimation(
      parent: animation,
      curve: Curves.linear,
    );

    final radius = radiusTween.evaluate(easeInAnimation);
    final offset = offsetTween.evaluate(animation);
    final size = sizeTween.evaluate(easeInAnimation);

    Widget positionedClippedChild(Widget child) => Positioned(
          width: size!.width,
          height: size.height,
          left: offset.dx,
          top: offset.dy,
          child: ClipRRect(
            borderRadius: radius,
            child: child,
          ),
        );

    return Stack(
      children: [
        positionedClippedChild(Container(color: _buttonColor)),
      ],
    );
  }

  _moveToNextPage({
    required BuildContext context,
    required Widget page,
    bool stopAnimation = false,
    required NavigationCallback callback,
  }) {
    final RenderBox buttonRenderBox =
        _buttonKey.currentContext!.findRenderObject() as RenderBox;
    final buttonSize = buttonRenderBox.size;
    final buttonOffset = buttonRenderBox.localToGlobal(Offset.zero);

    final pageBuilder = PageRouteBuilder(
      transitionDuration: widget.transitionDuration,
      pageBuilder: (BuildContext context, Animation<double> animation,
              Animation<double> secondaryAnimation) =>
          page,
      transitionsBuilder: (BuildContext context, Animation<double> animation,
              Animation<double> secondaryAnimation, Widget child) =>
          _buildTransition(context, child, animation, buttonSize, buttonOffset),
    );

    callback(pageBuilder);

    if (stopAnimation) _stopAnimation();
  }
}

typedef NavigationCallback = Function(PageRouteBuilder);

/// This controller defines a set of helper methods
/// that notifies the loading button about the state of the
/// animation
class LoadingButtonController {
  late _LoadginButtonState _state;

  /// Stops the loading animation and init the error state
  void onError() {
    _state._isError = true;
    _state._stopAnimation();
  }

  //new
  void onSuccess() async {
    _state._success();
    await Future.delayed(_state._durationSuccess);

    _state._stopAnimation();
  }

  /// Inits the loading state and fires [onSubmit] callback
  /// When the loading animation is running, the button will
  /// be disabled
  void startLoadingAnimation() {
    _state._iniAnimation();
  }

  /// Stops loading animations
  void stopLoadingAnimation() {
    _state._stopAnimation();
  }

  /// Inits the transition to the next page
  /// Please provide the next [page] that you want to present to the user.
  /// If you want to stop the button animation once the [page] is presented,
  /// you can set [stopAnimation] as true.
  /// Finally, provide a [navigationCallback] that will let you set the navigation
  /// behavior, this callback will return the new route.
  void moveToScreen({
    required BuildContext context,
    required Widget page,
    required stopAnimation,
    required NavigationCallback navigationCallback,
  }) {
    _state._moveToNextPage(
      context: context,
      page: page,
      stopAnimation: stopAnimation,
      callback: navigationCallback,
    );
  }
}


example/lib/main.dart
// ignore_for_file: avoid_print

import 'package:flutter/material.dart';
import 'package:font_awesome_flutter/font_awesome_flutter.dart';
import 'package:tbib_loading_transition_button_and_social/tbib_loading_transition_button_and_social.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Sign in Button Demo',
      theme: ThemeData.light(),
      darkTheme: ThemeData.dark(),
      themeMode: ThemeMode.system,
      home: const MyHomePage(title: 'Loading Button Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, this.title}) : super(key: key);
  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final String _buttonClick = "click sign in button";
  final _controllerSign = LoadingSignButtonController();
  final _controller = LoadingButtonController();
  bool signInAnimPage = false;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: Center(
        child: SingleChildScrollView(
          child: !signInAnimPage
              ? Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    LoadingButton(
                      controller: _controller,
                      color: Colors.blue,
                      errorColor: Colors.red,
                      child: const Text("Test"),
                      durationSuccess: const Duration(seconds: 1),
                      duration: const Duration(milliseconds: 500),
                      successWidget: const FaIcon(FontAwesomeIcons.checkCircle),
                      onSubmit: () {
                        print('onSubmit');
                      },
                    ),
                    const SizedBox(height: 5.0),
                    Row(
                      mainAxisSize: MainAxisSize.min,
                      children: [
                        ElevatedButton(
                          onPressed: () => _controller.startLoadingAnimation(),
                          child: const Text('Start'),
                        ),
                        const SizedBox(width: 8),
                        ElevatedButton(
                          onPressed: () => _controller.stopLoadingAnimation(),
                          child: const Text('Stop'),
                        ),
                        const SizedBox(width: 8),
                        ElevatedButton(
                          onPressed: () => _controller.onError(),
                          child: const Text('Error'),
                        ),
                        const SizedBox(width: 8),
                        ElevatedButton(
                          onPressed: () => _controller.onSuccess(),
                          child: const Text('success'),
                        ),
                      ],
                    ),
                    const SizedBox(height: 5.0),
                    ElevatedButton(
                      onPressed: () => setState(() {
                        signInAnimPage = true;
                      }),
                      child: const Text('Sign Button'),
                    ),
                  ],
                )
              : Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Text(
                      _buttonClick,
                      style: Theme.of(context).textTheme.headline4,
                    ),
                    const SizedBox(height: 5.0),
                    LoadingSignButton(
                      controller: _controllerSign,
                      buttonType: ButtonType.google,
                      progressIndicatorColor: Colors.grey,
                      durationSuccess: const Duration(seconds: 1),
                      duration: const Duration(milliseconds: 500),
                      successWidget: const FaIcon(FontAwesomeIcons.checkCircle),
                      onSubmit: () {
                        _controllerSign.startLoadingAnimation();
                        //  _buttonClick = "google";
                      },
                      fontSize: 20,
                      imageSize: 16,
                    ),
                    const SizedBox(height: 5.0),
                    LoadingSignButton.mini(
                      controller: _controllerSign,
                      buttonType: ButtonType.google,
                      progressIndicatorColor: Colors.grey,
                      durationSuccess: const Duration(seconds: 1),
                      duration: const Duration(milliseconds: 500),
                      successWidget: const FaIcon(FontAwesomeIcons.checkCircle),
                      onSubmit: () {},
                    ),
                    const SizedBox(
                      height: 5.0,
                    ),
                    Row(
                      mainAxisSize: MainAxisSize.min,
                      children: [
                        ElevatedButton(
                          onPressed: () =>
                              _controllerSign.startLoadingAnimation(),
                          child: const Text('Start'),
                        ),
                        const SizedBox(width: 8),
                        ElevatedButton(
                          onPressed: () =>
                              _controllerSign.stopLoadingAnimation(),
                          child: const Text('Stop'),
                        ),
                        const SizedBox(width: 8),
                        ElevatedButton(
                          onPressed: () => _controllerSign.onError(),
                          child: const Text('Error'),
                        ),
                        const SizedBox(width: 8),
                        ElevatedButton(
                          onPressed: () => _controllerSign.onSuccess(),
                          child: const Text('success'),
                        ),
                      ],
                    ),
                    const SizedBox(
                      height: 5.0,
                    ),
                    ElevatedButton(
                      onPressed: () => setState(() {
                        signInAnimPage = false;
                      }),
                      child: const Text('Normal Button'),
                    ),
                  ],
                ),
        ),
      ),
    );
  }
}
